linera-storage/src/lib.rs

```diff
@@ -358,28 +358,26 @@ where
         &self,
         description: &ApplicationDescription,
     ) -> Result<UserContractCode, ExecutionError> {
-        match self.user_contracts.entry(description.into()) {
-            Entry::Occupied(entry) => Ok(entry.get().clone()),
-            Entry::Vacant(entry) => {
-                let contract = self.storage.load_contract(description).await?;
-                entry.insert(contract.clone());
-                Ok(contract)
-            }
+        let application_id = description.into();
+        if let Some(contract) = self.user_contracts.get(&application_id) {
+            return Ok(contract.clone());
         }
+        let contract = self.storage.load_contract(description).await?;
```

Comment:
In case of race condition, we would load the contract twice and drop the first instance. Not sure if that's a problem. If not we may want to add a comment to help future readers.

cc @afck @MathieuDutSik